src/gc-stock.c

```diff
@@ -388,24 +388,16 @@ static void sweep_weak_refs(void)
         jl_ptls_t ptls2 = gc_all_tls_states[i];
         if (ptls2 != NULL) {
             size_t n = 0;
-            size_t ndel = 0;
+            size_t i = 0;
             size_t l = ptls2->gc_tls_common.heap.weak_refs.len;
             void **lst = ptls2->gc_tls_common.heap.weak_refs.items;
-            if (l == 0)
-                continue;
-            while (1) {
-                jl_weakref_t *wr = (jl_weakref_t*)lst[n];
+            // filter with preserving order
+            for (i = 0; i < l; i++) {
+                jl_weakref_t *wr = (jl_weakref_t*)lst[i];
                 if (gc_marked(jl_astaggedvalue(wr)->bits.gc))
-                    n++;
-                else
-                    ndel++;
-                if (n >= l - ndel)
-                    break;
-                void *tmp = lst[n];
-                lst[n] = lst[n + ndel];
-                lst[n + ndel] = tmp;
+                    lst[n++] = wr;
```

Comment:
I don't love `++` when combined with other things happening because it makes me think if it returns the previous or the next value